lib/src/middleware.dart

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'client.dart';
+import 'handler_client.dart';
+import 'request.dart';
+import 'response.dart';
+
+typedef Client Middleware(Client inner);
+
+typedef Future<Response> ResponseHandler(Response response);
+typedef Future<Request> RequestHandler(Request request);
+
+Middleware createMiddleware({
+  RequestHandler requestHandler,
+  ResponseHandler responseHandler,
+  CloseHandler closeHandler,
```

Comment:
I'd call this `onClose`, for consistency with `new Client.handler()` and other Dart APIs. It's also not *really* a handler, since it's not invoked for a request.